### PR TITLE
Warn against Data.Semigroup.Option and Data.Monoid.First/Last

### PR DESCRIPTION
**Pull request by andrewthad**
This is phase 1 of https://ghc.haskell.org/trac/ghc/ticket/15028.

cc @ekmett @RyanGlScott

### CODE REVIEW

**Comment by RyanGlScott** [comment]
Is this finished? It looks like this only adds docs for `First`, not `Last` or `Option`.

**Comment by andrewthad** [comment]
Sorry, wasn't finished yet. I just pushed a warning about `Data.Semigroup.Option`. I wanted to have someone read over the warning for `Data.Monoid.First` and make sure it sounded good. Once you or Edward confirm this, I'll copy the same warning into the haddocks for `Data.Monoid.Last`.

File: libraries/base/Data/Monoid.hs [hunk 1]
```diff
@@ -88,6 +88,16 @@ import Data.Semigroup.Internal
...
 -- >>> getFirst (First (Just "hello") <> First Nothing <> First (Just "world"))
 -- Just "hello"
+--
+-- Use of this type is discouraged. Note the following equivalence:
+--
+-- > Data.Monoid.First x === Maybe (Data.Semigroup.First x)
+--
+-- In additional to being equivalent in the structural sense, the two
+-- also have 'Monoid' instances that behave the same. This type will
+-- be marked deprecated in GHC 8.8. It will be removed in GHC 8.10.
```
Comment (RyanGlScott) [inline libraries/base/Data/Monoid.hs]: Maybe shorten this to "be marked as deprecated in GHC 8.8, and removed in GHC 8.10".

Comment (andrewthad) [reply libraries/base/Data/Monoid.hs]: Good call. Done.

**Comment by andrewthad** [comment]
I've made the suggested improvement to the wording. I've copied the warning to `Last` as well.

**Comment by ekmett** [comment]
Works for me.

**Comment by bgamari** [comment]
Well, if it works for Ed it works for me.

**Comment by bgamari** [comment]
Merged.
